### ISSUE TITLE
v3 - Modern PHP Framework

### ISSUE THREAD

**Issue by leevigraham** [FalconChristmas/fpp#493]
Hi @dkulp and @cpinkham,

I'm new to FPP but I have been a php developer for 15+ years developing CodeIgniter, Symfony and Javascript apps.

Looking over the source code of FPP it's clear that the project has grown organically for many years and is all custom code (no framework / documentation). 

It looks like the app could be rewritten as a Symfony (or Laravel) app with console commands. This could bring more developers like myself into the project. I understand this could be a very large undertaking.

I was wondering if there is any roadmap for a v3 that would bring the code inline with modern day PHP dev practices and frameworks?

Cheers Leevi

**Comment by cpinkham** [comment]
Thanks for the input.  Tim Debo started a FPP UI rewrite a couple years ago.  That code is in the new-ui branch, but it hasn't been rolled out.  It is something we were originally hoping to get into v2.0 when we wanted to release v2.0 back in 2017, but that wasn't possible mainly due to time constraints.  I don't think the branch has been touched much lately.

We definitely agree that the UI could be more modern and have been slowly working on adding support to the backend to allow for that.  FPP v2.x has a REST API built into fppd that is used for a few backend things already.  We also recently started adding a frontend API and some endpoints to make it easier for xLights to interact with FPP for manipulating whole playlists and other config files.

As you stated, replacing the UI would be a very large undertaking and require a lot of time.  FPP has basically been a one or two man operation with some help from others here and there along the way since it was created in 2013.  IMHO, the most important part of FPP is the backend and where we spend most of our time focused.  The thought was that if the lights don't blink when they should, then having a modern, pretty UI didn't help.  My concerns with a forklift UI upgrade have always been centered around the time it would take to make the rest of FPP work with that new UI and what dev time it would take away from the core of FPP.  For most of FPP's life, I have been the single core dev and therefore, replacing the UI has been low priority.  Unfortunately, I don't have the free time to spend coding FPP 15-20 hour per week like I used to in the beginning, so the things I do spend time on now are being prioritized a lot more than they used to be.

It would be nice if we could come up with a way to phase in a new UI or at least run in parallel so that we don't spend most of a year dev cycle adding in support for a new UI without getting any of the core desired features added for the next season.  Perhaps implementing the backend and frontend API's are the first step, then we can build the new UI on top of those or possibly update the old UI to use and burn in the API's.

Thoughts and input are welcome, but do you mind if we take it to the FPP forum instead of a github issue.  If you want to help work on this, we could also give you access to the FPP Developer subforum for discussion there.

**Comment by dkulp** [comment]
The other question is whether php is even the right option at this point.  It's quite possible that something like node.js would be more appropriate.  No idea really.  Web frontend's is also not by thing.  But that is an important point:  I'm not a web frontend person.  I'm not a PHP developer (I hadn't even touched PHP before working on FPP).  Someone more familiar with this space would need to drive any changes here.

As Chris said, this is something more appropriate as a discussion on the forums.

**Comment by leevigraham** [comment]
@cpinkham, 

Thanks for the reply.

> Tim Debo started a FPP UI rewrite a couple years ago. That code is in the new-ui branch, but it hasn't been rolled out. It is something we were originally hoping to get into v2.0 when we wanted to release v2.0 back in 2017, but that wasn't possible mainly due to time constraints. I don't think the branch has been touched much lately.

I'll check it out to see where Tim was heading.

> We definitely agree that the UI could be more modern and have been slowly working on adding support to the backend to allow for that. FPP v2.x has a REST API built into fppd that is used for a few backend things already. We also recently started adding a frontend API and some endpoints to make it easier for xLights to interact with FPP for manipulating whole playlists and other config files.

Can you explain what the backend is. I've looked through the repo and it seems there's a lot of python (could be perl) and PHP. An overview of the whole system would be good to know to see where the separation of the front end starts / stops.

> As you stated, replacing the UI would be a very large undertaking and require a lot of time. FPP has basically been a one or two man operation with some help from others here and there along the way since it was created in 2013.

Clearly it's a passion for everyone involved 😄 

> IMHO, the most important part of FPP is the backend and where we spend most of our time focused. The thought was that if the lights don't blink when they should, then having a modern, pretty UI didn't help.

As the poster in our office says: Make it valuable, then easy. then fast, then pretty.

> It would be nice if we could come up with a way to phase in a new UI or at least run in parallel so that we don't spend most of a year dev cycle adding in support for a new UI without getting any of the core desired features added for the next season. Perhaps implementing the backend and frontend API's are the first step, then we can build the new UI on top of those or possibly update the old UI to use and burn in the API's.

Yep, I wouldn't want core FPP development to suffer especially if there's a bunch of high priority features that need implementing.

> Thoughts and input are welcome, but do you mind if we take it to the FPP forum instead of a github issue. If you want to help work on this, we could also give you access to the FPP Developer subforum for discussion there.

Maybe I should have ready that first before replying above 😆.

------

Just saw  @dkulp  comment come through.

> The other question is whether php is even the right option at this point. It's quite possible that something like node.js would be more appropriate. 

That is something to consider. My main experience is with PHP on the backend and vuejs / react / jquery on the front end so I would lean that way. Again an overview of the front end / backend architecture would help trying to figure out a recommendation for future development.

Cheers Leevi

**Comment by cpinkham** [comment]
Replying in https://falconchristmas.com/forum/index.php/topic,10470 so we can move discussion there.

**Comment by leevigraham** [comment]
👍
